TombEngine/Objects/TR3/Trap/ElectricCleaner.cpp

```diff
@@ -0,0 +1,426 @@
+#include "Game/effects/tomb4fx.h"
+#include "Game/Lara/lara_helpers.h"
+//#include "Math/Math.h"
+//#include "Specific/level.h"
+#include "Specific/setup.h"
+
+using namespace TEN::Effects::Items;
+using namespace TEN::Effects::Spark;
+
+// ItemFlags[0]:		Rotation speed and heading angle.
+// ItemFlags[1]:		Flags, each bit is used to check the status of a flag
+//						b0: flagDoDetection
+//						b1: flagTurnRight
+//						b2: flagPriorityForward
+//						b3: flagAntiClockWiseOrder
+//						b4: flagStopAfterKill - If true the cleaner will stop when kills Lara.
+// ItemFlags[2]:		Movement velocity.
+// ItemFlags[3, 4, 5]:	Counters for dynamic lights and sparks.
+// ItemFlags[6]:		Goal direction angle.
+
+// OCB:
+// 0:			  Stop after killing the player.
+// Anything else: Don't stop after killing the player.
+
+namespace TEN::Entities::Traps
+{
+	constexpr auto ELECTRIC_CLEANER_VELOCITY  = BLOCK(1 / 16.0f);
+	constexpr auto ELECTRIC_CLEANER_TURN_RATE = 1024;
+
+	const auto ElectricCleanerHarmJoints = std::vector<unsigned int>{ 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13 };
+
+	std::vector <ItemInfo*> MyPushablesList = {};
+
+	void InitialiseElectricCleaner(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		// Align to the middle of the block.
+		item.Pose.Position.x = (item.Pose.Position.x & ~WALL_MASK) | (int)BLOCK(0.5f);
+		item.Pose.Position.z = (item.Pose.Position.z & ~WALL_MASK) | (int)BLOCK(0.5f);
+
+		// Init flags.
+		item.ItemFlags[0] = ELECTRIC_CLEANER_TURN_RATE;
+		item.ItemFlags[1] = 0;
+		item.ItemFlags[2] = ELECTRIC_CLEANER_VELOCITY;
+		item.ItemFlags[6] = item.Pose.Orientation.y;
+		item.Collidable = true;
+
+		if (item.TriggerFlags)
+			item.ItemFlags[1] &= ~(1 << 4);	// Turn off 1st bit for flagStopAfterKill.
+		else
+			item.ItemFlags[1] |= (1 << 4);	// Turn on 1st bit for flagStopAfterKill.
+
+		CollectLevelPushables(MyPushablesList);
+	}
+
+	void ElectricCleanerControl(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		auto& object = Objects[item.ObjectNumber];
+		
+		auto& rotationVel = item.ItemFlags[0];
+		auto& moveVel = item.ItemFlags[2];
+		auto& goalAngle = item.ItemFlags[6];
+
+		if (!TriggerActive(&item))
+		{
+			if (moveVel > 0)
+			{
+				moveVel = 0;
+				SoundEffect(SFX_TR3_CLEANER_FUSEBOX, &item.Pose);
+			}
+
+			return;
+		}
+
+		if (moveVel <= 0)
+			return;
+
+		float angleDifference = abs(TO_RAD(goalAngle) - TO_RAD(item.Pose.Orientation.y));
+
+		bool flagDoDetection		= ((item.ItemFlags[1] & (1 << 0)) != 0);
+		bool flagTurnRight			= ((item.ItemFlags[1] & (1 << 1)) != 0);
+		bool flagPriorityForward	= ((item.ItemFlags[1] & (1 << 2)) != 0);
+		bool flagAntiClockWiseOrder	= ((item.ItemFlags[1] & (1 << 3)) != 0);
+
+		auto col = GetCollision(item.Pose.Position.x, item.Pose.Position.y, item.Pose.Position.z, item.RoomNumber);
+
+		float yaw = TO_RAD(item.Pose.Orientation.y);
+
+		Vector3 ForwardDirection = Vector3(sin(yaw), 0, cos(yaw));
+		ForwardDirection.Normalize();
```

Comment:
Also make these `forwardDirection` and `rightDirection`, since they are local variables.

Reply:
Done